Test on on OS X using Travis

puremourning opened the pull request:
Fixes https://github.com/Valloric/ycmd/issues/191

Dependency
====

Merging this will require @Valloric to request OS X support for ycmd repo. This requires sending a mail to support@travis-ci.com and requesting to feature-flag the repo for multiple OS support.

Overview
====

This involved 2 changes:
 - switching to the new Travis infrastructure, and
 - getting build scripts set up to bootstrap on OS X

Unfortunately, the "python" Travis config is not yet supported on OS X, so we actually transition to the "csharp" one (because that just handles installing mono for us), and we implement the required multi-python-environment stuff manually.

This involves using a third-party repo for Linux and some cool built-in multi-lib stuff on OS X. 

Multiple Python Versions
========

We use an environment variable YCMD_PYTHON_VERSION in the travis config to set
the major.minor python version to test with.

On linux: We use the ppa fkrull/deadsnakes
(https://launchpad.net/~fkrull/+archive/ubuntu/deadsnakes) to install specific
major.minor python version.
On OSX: Apple ships both python 2.6 and 2.7 as part of the base Yosemite OS
release, as such we don't need to install anything. All we need to do is tell
the OS which "default" python version to use:
http://stackoverflow.com/questions/6998545/how-can-i-make-python-2-6-my-default-in-mac-os-x-lion

Annoyingly, however, Apple doesn't ship libpython2.6.dylib in the same
directory pattern as libpython2.7.dylib (it is only in /usr/lib rather than
also in /python_prefix/lib. A small change is required to build.py to support
this.

Timing
====

While the new infrastructure is faster for Linux builds, sadly the OS X builds take quite a long time, as there is significantly less resource available in Travis' cloud. This makes the whole run take about 1 hour, rather than the previous 20 minutes or so.

Other changes
====

* Fixed a bug where we were using `check_output` on python 2.6. Simply use a backport of it.
* `ordereddict` is not installed by default on python 2.6, so we add it to `test_requirements.txt`

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/240)
<!-- Reviewable:end -->

puremourning commented:
OK it seems the multiple OS support either doesn't require the flag, or it is inherited somehow, because the build is currently running on Valloric/ycmd in travis.

vheon commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[.travis.yml, line 20 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0NgER0TBNfTtHZOjD)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/.travis.yml#L20)):</sup>
Have you thought about using a matrix here? could make this more readable, something like:

```
env:
  global:
    - MONO_THREADS_PER_CPU=2000
    - MONO_MANAGED_WATCHER=disabled
  matrix:
    - USE_CLANG_COMPLETER="true"
    - USE_CLANG_COMPLETER="false"
    - YCMD_PYTHON_VERSION="2.7"
    - YCMD_PYTHON_VERSION="2.6"
```

what do you think? (I didn't test it)

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[.travis.yml, line 20 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0NgER0TBNfTtHZOjD)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/.travis.yml#L20)):</sup>
I think I did try that at one point. I can look into it. Depends how strongly people feel about it. I don't suppose it will change often and the net result is the same. 

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

<sup>**[.travis.yml, line 20 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0NgER0TBNfTtHZOjD)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/.travis.yml#L20)):</sup>
It was just a point for readability: when you have four env variable is difficult to see what change from one line to the other, at least for me, that's all; I don't suppose it will change any time soon either so I think that, yes is a matter of taste.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

Valloric commented:
Holy hell, that looks like it was a ton of work. Many thanks!

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

<sup>**[.travis.yml, line 20 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0NgER0TBNfTtHZOjD)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/.travis.yml#L20)):</sup>
I kinda agree with @vheon here; this is substantially more readable (and thus maintainable).

---

<sup>**[travis_install.osx.sh, line 19 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-1FhdvahxuIjpEY9Ek)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/travis_install.osx.sh#L19)):</sup>
A stray `#` char appears! :)

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

puremourning commented:
You know, the total amount of "work" wasn't that great - and actually went down as I refined the solution over a few iterations. Getting the python stuff to work was quite a journey. The tricky part was that it was so time-consuming :). Testing was painful, because of the change, push, wait, fail cycle you get with any type of CI testing. I eventually set up a VM in parallels and wrote a 'travis-fake' environment script to test iterations on the scripts. Also, I fixed and then unfixed after @micbou kindly fixed otherwise, a load of test failures. I'm happier with this because it is more self-contained. Anyway, I did most of the work when I was on holiday, so it was a nice distraction.

---

Review status: 4 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

<sup>**[.travis.yml, line 20 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0NgER0TBNfTtHZOjD)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/.travis.yml#L20)):</sup>
OK sure. Changed to use `global` section for the invariant env vars. The rest had to be as it was though to get the right combinations actually run.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

micbou commented:
General suggestions:
 - Move the Travis scripts in a new folder `travis`. Rename the files `travis_install.sh`, `travis_install.osx.sh`, and `travis_install.linux.sh` respectively to `install.sh`, `osx.sh`, and `linux.sh`.
 - Do not use curly braces for environment variables. Be consistent.
 - Start lines in comment with a capital letter.

I think there is some issues with the python version reported by CMake:
 - on Linux, when `YCMD_PYTHON_VERSION=2.6`, PythonLibs is python 2.7. This was already the case before.
 - on OS X, when `YCMD_PYTHON_VERSION=2.7`, PythonInterp is not the python from virtualenv and python 2.6 is used.

---

Review status: 4 of 6 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

<sup>**[build.py, line 68 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-13BW35raokvczIstC)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/build.py#L68)):</sup>
I would put this comment just above the function.

---

<sup>**[build.py, line 80 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-17FRFzd-_svvtr6oJ)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/build.py#L80)):</sup>
Update coding style in this function.

---

<sup>**[build.py, line 87 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0v_vdq0wrmm4uIud-)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/build.py#L87)):</sup>
Use parentheses for Python 3 compatibility and commas instead of quotation marks when possible.

---

<sup>**[build.py, line 97 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0vlcg1uH-SeAaU3ke)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/build.py#L97)):</sup>
Use parentheses and commas.

---

<sup>**[build.py, line 106 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0vnnpUMlckHoVe01s)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/build.py#L106)):</sup>
Use parentheses and commas.

---

<sup>**[build.py, line 122 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0vucGkPA_NXR4QHtL)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/build.py#L122)):</sup>
Merge those lines with ```sys.exit( 'ERROR: Unable to find an appropriate python library' )```

---

<sup>**[build.py, line 127 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0wKxbV1PhdAwlA9zA)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/build.py#L127)):</sup>
Use parentheses and commas.

---

<sup>**[travis_install.osx.sh, line 10 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-13o4_ZVY1wceELWOw)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/travis_install.osx.sh#L10)):</sup>
Typo.

---

<sup>**[travis_install.osx.sh, line 15 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-390JJ2y9aYICZ1svl)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/travis_install.osx.sh#L15)):</sup>
Use the SO short link.

---

<sup>**[travis_install.sh, line 8 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-1-peKjFF1spfVDB7M)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/travis_install.sh#L8)):</sup>
I don't see the point. If this file does not exist, the sourcing will fail anyway.

---

<sup>**[travis_install.sh, line 38 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0xmqMvG4d4Y3dGz8F)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/travis_install.sh#L38)):</sup>
This should be in the `env.global` Travis setting.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

puremourning commented:
> Do not use curly braces for environment variables

Any particular reason? I use a lot of scripting languages and I find that the addition of `{` and `}` around variables in certain scenarios is useful for showing the boundaries without worrying about particular subleties of the language at hand, such as when used as part of a path 

```bash
XYZ=/blah/${blah}/:${PATH}:/blah/blah${TEST}
```

I'm happy to make any changes, just want to understand the rationale, in case I am missing some subtlety.

> Start lines in comment with a capital letter

I thought I had. I will check. There are occasions where I have started with phases like 

```
# virtualenv does not....
```

The lower-case `v` there is intentional, as I'm referring to the `virtualenv` command.

> on Linux, when YCMD_PYTHON_VERSION=2.6, PythonLibs is python 2.7. This was already the case before.

Quite, and I think this might be some sort of bug. 

> on OS X, when YCMD_PYTHON_VERSION=2.7, PythonInterp is not the python from virtualenv and python 2.6 is used

I realised this, but I suspect that is the broken cmake talking. The tests themselves, I believe, do indeed run with the correct interpreter. That's why I put the check in `travis_install.sh`.

---

Review status: 4 of 6 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

<sup>**[build.py, line 87 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0v_vdq0wrmm4uIud-)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/build.py#L87)):</sup>
Commas `,`? Do you mean like:

```python
print( ,this is between 2 commas, )
```

Or are you referring to `'` single-quote? I didn't realise there was a distinction between `'` and `"` in Python.

```python
print( 'this is between 2 single quotes' )
print( "this is between 2 double quotes" )
```

According to [the docs](https://docs.python.org/3.5/reference/lexical_analysis.html) there is no difference:

```
shortstring     ::=  "'" shortstringitem* "'" | '"' shortstringitem* '"'
```

Likely I'm missing something subtle. Would you mind educating me :) ?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

micbou commented:
> Any particular reason? I use a lot of scripting languages and I find that the addition of { and } around variables in certain scenarios is useful for showing the boundaries without worrying about particular subleties of the language at hand, such as when used as part of a path

> XYZ=/blah/${blah}/:${PATH}:/blah/blah${TEST}

> I'm happy to make any changes, just want to understand the rationale, in case I am missing some subtlety.

Forget what I said. Without curly braces in paths, the syntax coloration will be messed up in some cases.
For consistency, we should maybe always use curly braces even if not in a path. For example, curly braces are used here:
```
defaults write com.apple.versioner.python Version ${YCMD_PYTHON_VERSION}
```
but not here:
```
python_version=$(python -c 'import sys; print "{0}.{1}".format( sys.version_info[0], sys.version_info[1] )')
echo "Checking python version (actual $python_version vs expected $YCMD_PYTHON_VERSION)"
test $python_version == $YCMD_PYTHON_VERSION
```
> I thought I had. I will check. There are occasions where I have started with phases like

>   virtualenv does not....

> The lower-case v there is intentional, as I'm referring to the virtualenv command.

I looked again and only found one missing capital letter that I highlighted.  I thought there were more.



>>    on Linux, when YCMD_PYTHON_VERSION=2.6, PythonLibs is python 2.7. This was already the case before.

> Quite, and I think this might be some sort of bug.

>> on OS X, when YCMD_PYTHON_VERSION=2.7, PythonInterp is not the python from virtualenv and python 2.6 is used

> I realised this, but I suspect that is the broken cmake talking. The tests themselves, I believe, do indeed run with the correct interpreter. That's why I put the check in travis_install.sh.

I am wondering if the ycmd libraries are linked to the correct python version.

---

Review status: 4 of 6 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

<sup>**[.travis.yml, line 24 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-3Uc0LDbX9od3iGjQp)** ([raw file](https://github.com/valloric/ycmd/blob/d626da27dc3303c48c23734efb2b8658cf9f6635/.travis.yml#L24)):</sup>
You can safely remove the quotation marks.

---

<sup>**[build.py, line 87 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0v_vdq0wrmm4uIud-)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/build.py#L87)):</sup>
Sorry, I meant single quotes. I mixed up the terms. This is only for consistency. Single quotes was used in the rest of the file.

---

<sup>**[build.py, line 97 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0vlcg1uH-SeAaU3ke)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/build.py#L97)):</sup>
Single quotes, not commas.

---

<sup>**[build.py, line 106 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0vnnpUMlckHoVe01s)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/build.py#L106)):</sup>
Single quotes, not commas.

---

<sup>**[build.py, line 127 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0wKxbV1PhdAwlA9zA)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/build.py#L127)):</sup>
Single quotes, not commas.

---

<sup>**[travis_install.osx.sh, line 3 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-3zJAlvxNqP4uey6gI)** ([raw file](https://github.com/valloric/ycmd/blob/d626da27dc3303c48c23734efb2b8658cf9f6635/travis_install.osx.sh#L3)):</sup>
Missing capital letter to run.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

puremourning commented:
I have updated the var usages for consistency.

For OS X, the correct python libs are used, the `CMakeLists.txt` do not reference `PYTHON_EXECUTABLE` which is what is reported by the line about `PythonInterp`. However, I suspect this means we are (and have always been) testing with libpython2.7 on Linux. I think fixing that is out of scope for this PR. If I can solve it, I will, but it is likely to be soul-destroyingly-time-consuming.

I have moved the scripts to a `travis` directory as suggested, but I have not renamed them as I believe `travis_install*.sh` is less likely to be accidentally run by a user. In any case, I have also added a small `readme.txt` to the directory.

---

Review status: 1 of 7 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

<sup>**[.travis.yml, line 24 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-3Uc0LDbX9od3iGjQp)** ([raw file](https://github.com/valloric/ycmd/blob/d626da27dc3303c48c23734efb2b8658cf9f6635/.travis.yml#L24)):</sup>
OK, done. Shrug :)

---

<sup>**[build.py, line 80 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-17FRFzd-_svvtr6oJ)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/build.py#L80)):</sup>
I didn't do this because apparently it is taken directly (unmodified) from the core lib in python 2.7. But I take the point and have made the changes.

---

<sup>**[build.py, line 87 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-0v_vdq0wrmm4uIud-)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/build.py#L87)):</sup>
I have replaced all non-essential uses of the evil double-quote, and changed all `print` statements to `print()` statements

---

<sup>**[travis_install.sh, line 8 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-1-peKjFF1spfVDB7M)** ([raw file](https://github.com/valloric/ycmd/blob/88d80eb1195fc2cadbe45f8af1a4bf9b2eb84c95/travis_install.sh#L8)):</sup>
Just makes it more obvious why it failed. Removed.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 8 of 9 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

vheon commented:
This LGTM

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

puremourning commented:
Appveyor still seems quite flaky. Don't think the failure is anything to do with this change.

vheon commented:
What failure? I see all green here.

puremourning commented:
I'm sure there was a problem. Maybe it retried? Shrug.

mispencer commented:
<sup>**[travis/travis_install.sh, line 23 \[r6\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-JWzEb7n5rCf7oE7fn)** ([raw file](https://github.com/valloric/ycmd/blob/aae0dc4201c8fd93c06713ab43b6e07c6257356e/travis/travis_install.sh#L23)):</sup>
serious => series 
<!-- Reviewable comment -K-JWzEb7n5rCf7oE7fo:-4222005 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

Valloric commented:
Reviewed 6 of 9 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

micbou commented:
You are certainly talking about the failure in the previous revision. It was a network issue during the submodules checkout. If it happens again, I will try to fix it.

Except a minor comment, I think this PR is ready to be merged.
<!-- Reviewable comment -K-Ysgxd1cDuCDWGUpFs:1317903980 -->

---

Reviewed 1 of 9 files at r5, 1 of 2 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[travis/readme.txt, line 2 \[r6\]](https://reviewable.io:443/reviews/valloric/ycmd/240#-K-YwyHucp_UM_z4bCIO)** ([raw file](https://github.com/valloric/ycmd/blob/aae0dc4201c8fd93c06713ab43b6e07c6257356e/travis/readme.txt#L2)):</sup>
I would rename this file `README.md` like in the `examples` folder.
<!-- Reviewable comment -K-YwyHucp_UM_z4bCIP:242239948 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

puremourning commented:
Thanks, I've updated for the minor comments.
<!-- Reviewable comment -K-_FsjixzHhuyzueuo9:-458023085 -->

---

Review status: 5 of 7 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

vheon commented:
LGTM
<!-- Reviewable comment -K-_em4sdfZYwsbSL90S:676834233 -->

---

Reviewed 2 of 3 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 1 of 3 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

Valloric commented:
Thanks!

@homu r+
<!-- Reviewable comment -K-a9H2y6lheofx6ctTs:-388066883 -->

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/240)
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 43195e1 has been approved by `Valloric`

<!-- @homu r=Valloric 43195e1d5b0dde5c4b94f150ec0d94473808619f -->

homu commented:
:hourglass: Testing commit 43195e1 with merge 8d6cf88...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/Valloric/ycmd/builds/83229780)
